server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/utilities/AvroUtilities.java

```diff
@@ -0,0 +1,287 @@
+            }
+        }
+        return fileReader.getSchema();
+    }
+
+    private static Schema readSchemaFromAvroDataSource(Configuration configuration, String dataSource) throws IOException {
+        DatumReader<GenericRecord> datumReader = new GenericDatumReader<>();
+        FsInput inStream = new FsInput(new Path(dataSource), configuration);
+
+        try (DataFileReader<GenericRecord> fileReader = new DataFileReader<>(inStream, datumReader)) {
+            return fileReader.getSchema();
+        }
+    }
+
+    /*
+     * if user provided a full path, use that.
+     * otherwise we need to check classpath
+     */
+    private static File searchForFile(String schemaName) throws UnsupportedEncodingException {
+        File file = new File(schemaName);
+        if (!file.exists()) {
+            URL url = AvroUtilities.class.getClassLoader().getResource(schemaName);
+
+            /** Testing that the schema resource exists. */
+            if (url == null) {
+                return null;
+            }
+            file = new File(URLDecoder.decode(url.getPath(), "UTF-8"));
+        }
+        return file;
+    }
+
+    private static Schema generateSchema(List<ColumnDescriptor> tupleDescription) throws IOException {
+        String colName;
+        int colType;
+
+        Schema schema = Schema.createRecord("tableName", "", COMMON_NAMESPACE, false);
+        List<Schema.Field> fields = new ArrayList<>();
+
+        for (ColumnDescriptor cd : tupleDescription) {
+            colName = cd.columnName();
+            colType = cd.columnTypeCode();
+            // String delim = context.getOption("delimiter");
+            // columnDelimiter = delim == null ? ',' : delim.charAt(0);
+            fields.add(new Schema.Field(colName, getFieldSchema(DataType.get(colType), false, 1), "", null));
+        }
+
+        schema.setFields(fields);
+
+        return schema;
+    }
+
+    private static Schema getFieldSchema(DataType type, boolean notNull, int dim) throws IOException {
+        List<Schema> unionList = new ArrayList<>();
+        // in this version of gpdb, external table should not set 'notnull' attribute
+        unionList.add(Schema.create(Schema.Type.NULL));
+
+        switch (type) {
+            case BOOLEAN:
+                unionList.add(Schema.create(Schema.Type.BOOLEAN));
+                break;
+            case BYTEA:
+                unionList.add(Schema.create(Schema.Type.BYTES));
+                break;
+            case BIGINT:
+                unionList.add(Schema.create(Schema.Type.LONG));
+                break;
+            case SMALLINT:
+            case INTEGER:
+                unionList.add(Schema.create(Schema.Type.INT));
+                break;
+            case REAL:
+                unionList.add(Schema.create(Schema.Type.FLOAT));
+                break;
+            case FLOAT8:
+                unionList.add(Schema.create(Schema.Type.DOUBLE));
+                break;
+            case VARCHAR:
```

Comment:
why all the cases, if default is doing string too?
```suggestion
            default:
                unionList.add(Schema.create(Schema.Type.STRING));
```

Reply:
Yeah I reduced it down. Thanks.